xskillscore/core/probabilistic.py

```diff
@@ -352,3 +358,148 @@ def threshold_brier_score(
         else:
             return res.mean(dim)
+
+
+def rank_histogram(observations, forecasts, dim=None, member_dim='member'):
+    """Returns the rank histogram (Talagrand diagram) along the specified dimensions.
+
+        Parameters
+        ----------
+        observations : xarray.Dataset or xarray.DataArray
+            The observations or set of observations.
+        forecasts : xarray.Dataset or xarray.DataArray
+            Forecast with required member dimension ``member_dim``.
+        dim : str or list of str, optional
+            Dimension(s) over which to compute the histogram of ranks.
+            Defaults to None meaning compute over all dimensions
+        member_dim : str, optional
+            Name of ensemble member dimension. By default, 'member'.
+
+        Returns
+        -------
+        rank_histogram : xarray.Dataset or xarray.DataArray
+            New object containing the histogram of ranks
+
+        Examples
+        --------
+        >>> observations = xr.DataArray(np.random.normal(size=(3,3)),
+        ...                             coords=[('x', np.arange(3)),
+        ...                                     ('y', np.arange(3))])
+        >>> forecasts = xr.DataArray(np.random.normal(size=(3,3,3)),
+        ...                          coords=[('x', np.arange(3)),
+        ...                                  ('y', np.arange(3)),
+        ...                                  ('member', np.arange(3))])
+        >>> rank_histogram(observations, forecasts, dim='x')
+        <xarray.DataArray 'histogram_rank' (y: 3, rank: 4)>
+        array([[0, 1, 1, 1],
+               [0, 1, 0, 2],
+               [1, 0, 1, 1]])
+        Coordinates:
+          * y        (y) int64 0 1 2
+          * rank     (rank) float64 1.0 2.0 3.0 4.0
+
+        Notes
+        -----
+        See http://www.cawcr.gov.au/projects/verification/
+    """
+
+    def _rank_first(x, y):
+        """ Concatenates x and y and returns the rank of the first element along the last axes """
+        xy = np.concatenate((x[..., np.newaxis], y), axis=-1)
+        return bn.nanrankdata(xy, axis=-1)[..., 0]
+
+    if dim is not None:
+        if len(dim) == 0:
+            raise ValueError(
+                'At least one dimension must be supplied to compute rank histogram over'
+            )
+        if member_dim in dim:
+            raise ValueError(f'"{member_dim}" cannot be specified as an input to dim')
+
+    ranks = xr.apply_ufunc(
+        _rank_first,
+        observations,
+        forecasts,
+        input_core_dims=[[], [member_dim]],
+        dask='parallelized',
+        output_dtypes=[int],
+    )
+
+    bin_edges = np.arange(0.5, len(forecasts[member_dim]) + 2)
+    return histogram(
+        ranks, bins=[bin_edges], bin_names=['rank'], dim=dim, bin_dim_suffix=''
+    )
+
+
+def discrimination(
+    observations,
+    forecasts,
+    dim=None,
+    probability_bin_edges=np.linspace(-1 / 8, 1 + 1 / 8, 6),
```

Comment:
sorry @dougiesquire I wrote these comments earlier but there were not released somehow.

Reply:
I've since changed these after realising that `xhistogram` does something slightly different with `bins` than `numpy.histogram` (i.e. does **not** include the right edge of the right-most bin). However I still have a default value in my latest commit: `probability_bin_edges=np.linspace(0, 1 + 1e-8, 6)`. The `+ 1e-8` is needed to make sure that 1.0 is included in the last bin - this seems to be standard practice in `sklearn`, for example.

To answer your question, the key difference between the edges specified for contingency and those specified here is that the former are completely unbounded (they just specify how we want to categorise our data), whereas `probability_bin_edges` should always bounded between 0 and 1. I've chose the same default values as the `sklearn` implementations of these types of metric. So yes, I think it makes sense to have defaults specified here and not in Contingency.